### PR TITLE
Discussion 1 1

### PR DESCRIPTION
**Pull request by yongzx**

### CODE REVIEW

**Comment by adamist521** [comment]
Just integrated to slack channel!

File: pytorch/train.py [hunk 1]
```diff
@@ -194,6 +194,10 @@
     device=device, ext_len=args.ext_len)
 
 # adaptive softmax / embedding
+# Do I understand it correctly that cutoffs and tie_projs are embeddings?
```
Comment (adamist521) [inline pytorch/train.py]: `cutoffs` and `tie_projs` are some parameter on treating the data.
Embedding is created with `AdaptiveEmbedding` which is called inside `MemTransformerML`
https://github.com/rhditemo/transformer-xl/blob/a3682b4f0b62a9dc3cec043b85581986837eeea2/pytorch/train.py#L280
https://github.com/rhditemo/transformer-xl/blob/master/pytorch/mem_transformer.py#L433

File: pytorch/mem_transformer.py [hunk 2]
```diff
@@ -384,7 +384,7 @@ def __init__(self, n_head, d_model, d_head, d_inner, dropout, **kwargs):
                                      pre_lnorm=kwargs.get('pre_lnorm'))
 
     def forward(self, dec_inp, dec_attn_mask=None, mems=None):
-
+        # Is the class MultiHeadAttn passed to the next layer? Does PyTorch allows passing of class across layers?
```
Comment (adamist521) [inline pytorch/mem_transformer.py]: Not sure about that... have to look further in Pytorch stuff...

File: pytorch/mem_transformer.py [hunk 3]
```diff
@@ -579,6 +579,7 @@ def __init__(self, n_token, n_layer, n_head, d_model, d_head, d_inner,
         self._create_params()
 
     def backward_compatible(self):
+        # what is this function for?
```
Comment (adamist521) [inline pytorch/mem_transformer.py]: Haven't figured out completely but should be related to eval phase and training phase due to code...

https://github.com/rhditemo/transformer-xl/blob/a3682b4f0b62a9dc3cec043b85581986837eeea2/pytorch/eval.py#L61
https://github.com/rhditemo/transformer-xl/blob/a3682b4f0b62a9dc3cec043b85581986837eeea2/pytorch/train.py#L128

Comment (fullflu) [reply pytorch/mem_transformer.py]: I confirmed that if sample_softmax > 0;
```
* sparse optimizer and sparse scheduler is used in training steps
* LogUniformSampler is used in forward process
```

LogUniformSampler seems to be control negative sampling when training distributional representations.
Negative sampling should not be used in evaluation, so backward_compatible would be called in `eval.py` .

I have not checked the details of how the negative sampling is used and what the sparse optimizer/scheduler means.
